### PR TITLE
feat(SNI) Bun.serve/tls

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?
Fix https://github.com/oven-sh/bun/issues/9753 when using serverName:

```ts
const server = Bun.serve({
  development: true,
  async fetch(request: Request): Promise<Response> {
    // code here...
  },
  hostname: '0.0.0.0', //accessible hostname or ip to bind
  lowMemoryMode: false,
  port: 443,
  tls: {
    cert: Bun.file('./config/bun/localhost.pem'),
    key: Bun.file('./config/bun/localhost-key.pem'),
    serverName: 'get.babbel.test' // server name
  },
});
```
Fix https://github.com/oven-sh/bun/issues/6529 by adding support to multiple servernames:
 ```ts
 const server = Bun.serve({
  async fetch(request: Request): Promise<Response> {
    // code here...
  },
  hostname: '0.0.0.0', //accessible hostname or ip to bind
  port: 443,
  tls: [
   {
      cert: Bun.file('./config/bun/example.pem'),
      key: Bun.file('./config/bun/example-key.pem'),
      serverName: '*.example.com' // server name
    },
   {
      cert: Bun.file('./config/bun/test.pem'),
      key: Bun.file('./config/bun/test-key.pem'),
      serverName: '*.test.com' // server name
    }
  ],
});
```
And also fix https://github.com/oven-sh/bun/issues/8910 adding support to tls.Server.addContext https://nodejs.org/api/tls.html#serveraddcontexthostname-context

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I wrote automated tests

### CODE REVIEW

File: src/bun.js/api/bun/socket.zig [hunk 1]
```diff
@@ -814,18 +814,40 @@ pub const Listener = struct {
...
+    pub fn addServerName(this: *Listener, global: *JSC.JSGlobalObject, hostname: JSValue, tls: JSValue) JSValue {
+        if (!this.ssl) {
+            global.throwInvalidArguments("addServerName requires SSL support", .{});
+            return .zero;
+        }
+        if (!hostname.isString()) {
+            global.throwInvalidArguments("hostname pattern expects a string", .{});
+            return .zero;
+        }
+        const host_str = hostname.toSliceZ(
```
Comment (Jarred-Sumner) [inline src/bun.js/api/bun/socket.zig]: The poorly-named `toSliceZ`does not actually make it `[:0]const u8`

https://github.com/oven-sh/bun/blob/c5270f812105133975f343c4618c2411f5d899f3/src/bun.js/bindings/bindings.zig#L804-L821

File: src/bun.js/bindings/bindings.zig [hunk 2]
```diff
@@ -804,8 +804,8 @@ pub const ZigString = extern struct {
     pub fn toSliceZ(this: ZigString, allocator: std.mem.Allocator) Slice {
         if (this.len == 0)
             return Slice.empty;
-
-        if (is16Bit(&this)) {
+        // if is 16 bit or dont has a null terminator
+        if (is16Bit(&this) or this._unsafe_ptr_do_not_use[this.len] != 0) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.zig]: We cannot do this. It is an index out of bounds.

File: src/js/node/tls.ts [hunk 3]
```diff
@@ -1,7 +1,10 @@
 // Hardcoded module "node:tls"
 const { isArrayBufferView, isTypedArray } = require("node:util/types");
 const net = require("node:net");
-const { Server: NetServer, [Symbol.for("::bunternal::")]: InternalTCPSocket } = net;
+const {
+  Server: NetServer,
+  [Symbol.for("::bunternal::")]: [InternalTCPSocket, addServerName],
```
Comment (Jarred-Sumner) [inline src/js/node/tls.ts]: We can use a shared export for these now. No need to use this symbol anymore.

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
A few comments

File: src/js/node/tls.ts (hunk 3 above)
Comment (cirospaciari) [reply src/js/node/tls.ts]: interesting, do you have any example of this? did not find about shared exports

Comment (Jarred-Sumner) [reply src/js/node/tls.ts]: you could add `internal/net.ts` which would export `Symbol("kSharedInternal")`

Comment (cirospaciari) [reply src/js/node/tls.ts]: nice add $zig addServerName into internals and will do a followup with a refactor removing existing unnecessary symbols

**Comment by Jarred-Sumner** [comment]
yay

**Pull request merged**